homeassistant/components/rpi_gpio/binary_sensor.py

```diff
@@ -62,17 +76,43 @@ def __init__(self, name, port, pull_mode, bouncetime, invert_logic):
 
         rpi_gpio.setup_input(self._port, self._pull_mode)
 
+        # detection mode:
+        #   input: read GPIO input after any edge without polling
+        #   input_poll: read GPIO input after any edge with polling
+        #   edge: trigger on rising/falling edge and use polling to reset
+        if detect is DetectEnum.INPUT:
+            self._input = True
+            self._poll = False
+            self._edge = "BOTH"
+        elif detect is DetectEnum.INPUT_POLL:
+            self._input = True
+            self._poll = True
+            self._edge = "BOTH"
+        elif detect is DetectEnum.EDGE:
+            self._input = False
+            self._poll = True
+            self._edge = "FALLING" if self._invert_logic else "RISING"
```

Comment:
I'm not a fan of that code part. I think that is not needed and we need only store the `self._detect = detect`. There are 2 part of code where we need an decision based on this variable

Reply:
It is not nice to look at, but it takes the decision-making out of the time-critical bit of code. I would actually prefer `read_gpio` to be just `self._state = (rpi_gpio.read_input(self._port) ^ self._invert_logic) || ~self._input` (which reads the input first thing in all cases) but that is not very straightforward to read.